bug 1274413: json_verify.py dnt

groovecoder opened the pull request.

groovecoder commented:
Assigning to @fmarier.

fmarier commented on scripts/json_verify.py:
```diff
@@ -132,10 +132,8 @@ def reset():
-        for f in files:
-            if ".json" in f:
-                run(f)
+    for f in glob.glob("*.json"):
```
Much nicer!

fmarier commented on scripts/json_verify.py:
```diff
@@ -100,6 +100,13 @@ def find_uris(categories_json):
                 assert type(entity_json) is DictType
+                # pop dnt out of the dict, so we can iteritems() over the rest
+                try:
+                    dnt_value = entity_json.pop('dnt', '')
```
I assume that if `dnt` is missing, it returns the empty string here?

groovecoder replied:
Yes, `.pop('dnt', '')` returns the `dnt` item out of `entity_json` or defaults to `''`. So if the `dnt` item is missing, this returns an empty string.

fmarier commented:
r+, but please test a bunch of invalid JSON files in a different repo before merging. I want to make sure that we continue to see Travis errors if a PR breaks the list.

groovecoder commented:
Yeah, I agree. I don't want to get into a rabbit-hole, but I'm considering turning the script into an actual python test case and creating multiple JSON fixtures with lots of error cases.

groovecoder commented:
Okay, I only spent a **little** too much time trying to turn this into a python `unittest.TestCase` before I just gave the `json_verify.py` script a filename argument.

I'll start adding some invalid JSON files into a `test/` directory here to make it easy to test the script against invalid files. Do you have any good examples of invalid JSON files? Maybe some that include mistakes we've seen in the past?

groovecoder commented:
Okay, I added a known-valid and a known-invalid test fixture JSON file. I spot-checked that they pass/fail as expected. If you can spot-check, then I'm ready to merge this if you are.

But I could also look at running the script against all 3 files via tox and setting the known-invalid job as an allowed fail on travis ... ?

fmarier commented on tests/disconnect-blacklist-valid.json:
```diff
@@ -0,0 +1,121 @@
+        "Disconnect": [
+            {
+                "Facebook": {
+                    "dnt": "dnt",
```
This should be invalid. The allowed values are: `""`, `"eff"` and `"w3c"`.

groovecoder replied:
https://bugzilla.mozilla.org/show_bug.cgi?id=1274413#c0 says "eff", "dnt", or "".

fmarier commented:
Since you have a nice test harness for this now, I would suggest also adding a valid file but with a `"dnt": "bogus"` line (which makes it invalid).

groovecoder commented:
Added file with a bogus dnt value. Also updated README to include instructions for using the json_verify script and the test fixtures. I also protected the `master` branch here so that only pull requests/commits that have passed the Travis CI job can be merged into master.

groovecoder commented:
Still need to know if the `""`, `"eff"`, and `"w3c"` options are the valid ones, or if `""`, `"eff"`, and `"dnt"` options from the original bug description are the valid ones.

fmarier commented:
> Still need to know if the "", "eff", and "w3c" options are the valid ones, or if "", "eff", and "dnt" options from the original bug description are the valid ones.

Yes, that was a typo on bugzilla, sorry about that.

`dnt` is not valid. `eff`, `w3c` and the empty string are the valid ones.

groovecoder commented:
Updated script and test fixtures with "w3c" instead of "dnt"

fmarier commented on README.md:
```diff
@@ -1,24 +1,67 @@
 
-* disconnect-blacklist.json is a version controlled copy of the Disconnect.me public GPL anti-tracking list available here: https://services.disconnect.me/disconnect-plaintext.json
-* disconnect-entitylist.json is a transformed version of the public GPL list that Disconnect reformats for Firefox, in order to reduce the size of the entity list we send to Firefox via the Shavar service. 
+* `disconnect-blacklist.json` is a version controlled copy of the Disconnect.me public GPL anti-tracking list available here: [https://services.disconnect.me/disconnect-plaintext.json](https://services.disconnect.me/disconnect-plaintext.json)
```
You can also use `<http://example.com>` in markdown linkify URLs without having to repeat them.

fmarier commented:
r+
